### ISSUE TITLE
[Proposal] Grouped Const.

### ISSUE THREAD

**Issue by AdamSpeight2008** [dotnet/vblang#28]
[(Ported from Roslyn Repo)](https://github.com/dotnet/roslyn/issues/11165)

## Grouped Const

In the big open-source projects (I've seen) using C# or VB.net, tend to contain a lot of constants.
Eg [Roslyn Project System](https://github.com/dotnet/roslyn-project-system/blob/master/src/Common/IVsSqm.vb)

``` vb.net
        Public Const Undefined As UInt32 = 0
        Public Const DATAID_SQM_ACTIVATEDOCWINDOW As UInt32 = 1    ' (Stream) 
        Public Const DATAID_SQM_UICONTEXTCHANGED As UInt32 = 2    ' (Stream) 
        Public Const DATAID_SQM_PERFMEMORY As UInt32 = 3
        Public Const DATAID_SQM_STARTUPOS As UInt32 = 4
        Public Const DATAID_SQM_STARTUPOSVERSION As UInt32 = 5
        Public Const DATAID_SQM_STARTUPOSBUILD As UInt32 = 6
        Public Const DATAID_SQM_STARTUPOSSERVICEPACK As UInt32 = 7
        Public Const DATAID_SQM_STARTUPAPPID As UInt32 = 8
        Public Const DATAID_SQM_STARTUPAPPBUILDTYPE As UInt32 = 9
        Public Const DATAID_SQM_STARTUPAPPVERSION As UInt32 = 10
```

I think it would help if there was an alternative syntax for writing them. In c# and VB.net we have a nice syntax for writing `Enums` values. Why not have a similar syntax for constants.
## Proposed Syntaxes

**VB.net**

``` vb.net
Public Const UInt32
  Undefined                       =  0,
  DATAID_SQM_ACTIVATEDOCWINDOW    =  1, ' (Stream) 
  DATAID_SQM_UICONTEXTCHANGED     =  2, ' (Stream) 
  DATAID_SQM_PERFMEMORY           =  3,
  DATAID_SQM_STARTUPOS            =  4,
  DATAID_SQM_STARTUPOSVERSION     =  5,
  DATAID_SQM_STARTUPOSBUILD       =  6,
  DATAID_SQM_STARTUPOSSERVICEPACK =  7,
  DATAID_SQM_STARTUPAPPID         =  8,
  DATAID_SQM_STARTUPAPPBUILDTYPE  =  9,
  DATAID_SQM_STARTUPAPPVERSION    = 10
End Const
```

The compiler would consider this to be equivalently to as if it was the first.

**C#**

``` c#
public const uint32
{
  Undefined                       =  0,
  DATAID_SQM_ACTIVATEDOCWINDOW    =  1, // (Stream) 
  DATAID_SQM_UICONTEXTCHANGED     =  2, // (Stream) 
  DATAID_SQM_PERFMEMORY           =  3,
  DATAID_SQM_STARTUPOS            =  4,
  DATAID_SQM_STARTUPOSVERSION     =  5,
  DATAID_SQM_STARTUPOSBUILD       =  6,
  DATAID_SQM_STARTUPOSSERVICEPACK =  7,
  DATAID_SQM_STARTUPAPPID         =  8,
  DATAID_SQM_STARTUPAPPBUILDTYPE  =  9,
  DATAID_SQM_STARTUPAPPVERSION    = 10

}
```

**Caveats**
- All of the constants defined within a **grouped const** have the same type  
  - In this example `UInt32`
- All of the constants defined within a **grouped const** have the same visibility.
  - In this example: `Public`
- You can not change the type / visibility of an entry within a **grouped const**
  - eg `Private  DATAID_SQM_STARTUPOSSERVICEPACK =  7,` **error**

**Comment by Bill-McC** [comment]
or just use an enum ?   and/or allow enums to be of type string??

**Comment by AdamSpeight2008** [comment]
@Bill-McC  See original Roslyn Post, for rationale.

**Comment by Bill-McC** [comment]
@AdamSpeight2008 yep, read it. Didn't see anything compelling for VB, and as I noted you said, you would have used an enum, but for some "unknown reason" they had used const's. So rather than change the language for some unknown reason, we should question why const's were used when enums would have provided a much better experience not only in the declaring but also in the subsequent editing, use, and code maintenance stages.
Given the proposed constraint of a particular type, int32, et al, the proposed syntax would be identical to an enum in use in VB, given in VB you can Import the enum and can implicitly convert to the base type within strict semantics.

**Comment by AdamSpeight2008** [comment]
@Bill-McC  That's the point, they used `Const`. Thus we should make that experience better, which is the main aspect of this proposal.

**Comment by Bill-McC** [comment]
@AdamSpeight2008 okay, so a warning or refactoring suggesting to use an enum instead of a heap of constants of the same type

**Comment by scottdorman** [comment]
The only way using an enum would really work is if `String` is allowed as a base type for an enum __and__ an implicit conversion from enum to `String` exists (or really whatever the enum base type is).

As an example, in claims based identity, when creating a new claim, both the claim type and the claim value must be strings. 

Look at [corefx/ClaimTypes.cs](https://github.com/dotnet/corefx/blob/master/src/System.Security.Claims/src/System/Security/Claims/ClaimTypes.cs), and you'll see that all of the claims here are defined as `const` for that reason. That allows me to simply write `new Claim(ClaimTypes.NameIdentifier, user.Id)` without needing any `.ToString()` calls or casts.

Using a string-based enum (without the implicit conversion) I would have to write `new Claim(ClaimTypesEnum.NameIdentifier.ToString(), user.Id)` instead. With the implicit conversion, the code I have to write doesn't change.

It seems to me that this is all about not having to repeat the `Public Const As String` part of the declaration, allowing a more enum-like syntax to define them. The issue I see is that currently, enums are their own "top level" entity. They require a name and provide values. The proposed syntax doesn't allow that. I still have to define the constants in a class.  If we allowed some sort of `const` declaration that creates it's own "top level" entity, we end up with

```c#
public const ClaimTypes : string
{
   Expiration = "http://schemas.microsoft.com/ws/2008/06/identity/claims/expiration",
   Role = "http://schemas.microsoft.com/ws/2008/06/identity/claims/role",
   Anonymous = "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/anonymous",
   Country = "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/country",
   DateOfBirth = "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/dateofbirth",
   Email = "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress",
   Gender = "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/gender",
   GivenName = "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/givenname",
}
```

instead of

```c#
public static class ClaimTypes
{
   public const string Expiration = "http://schemas.microsoft.com/ws/2008/06/identity/claims/expiration";
   public const string Role = "http://schemas.microsoft.com/ws/2008/06/identity/claims/role";
   public const string Anonymous = "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/anonymous";
   public const string Country = "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/country";
   public const string DateOfBirth = "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/dateofbirth";
   public const string Email = "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress";
   public const string Gender = "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/gender";
   public const string GivenName = "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/givenname";
}
```

**Comment by Bill-McC** [comment]
@scottdorman Yes, today in VB.NET, there is an implicit cast from an enum to it's base type or wider.  You can't implicitly use an enum of int64 as an int32, but you can use it as an int64. So generally speaking the rule is an enum in vb is implicitly cast to it's base type. If there was a string enum type, calling ToString would be different from cast to string.
The question gets more complex around Flags as that would imply string(), not string. And an enum of Char with flags would imply string .
For cases other than string (and char), I think enums provide a good grouping and good editing experience and am loathe to see yet another way of declaring something introduced to the language when it is not a barrier type of issue. I really believe people struggle to remember all the different structures and syntax as is in the multi language universe many coexist in. The first thing I thought of when I saw the original example is "clippy" suggesting "it looks like you are trying to declare an enum".

For strings, I'm not sure there is enough reason to change enums to include them: I'd like to see more compelling cases. I don't see an issue with the current syntax, but I get folks might not like typing public const string line after line, although I would suggest a snippet or simply copy and paste ctrl+c, ctrl+v or ctrl+shift+v if copying other stuff as well

**Comment by AdamSpeight2008** [comment]
@Bill-McC  It's not only about writing the code, but reading it. Repeating redundant information just visually clutters code, especially since the `Public Const` are syntax highlighted. The red and blue parts stand out more than the identifier part,

**Comment by scottdorman** [comment]
@Bill-McC To be clear, I'm not suggesting that enums be changed to support string as a base type. (The proposal wasn't either.) There were comments both here and in the referenced proposal in the roslyn repo that suggested enums should be used instead.

The point is that declaring groups of constants (like in my example of claim types) is a pretty common scenario that could be made much simpler (both for typing and reading) by allowing an enum-like declaration of those constants. I proposed a slightly different syntax than what @AdamSpeight2008 originally had that makes the declaration more like an enum in syntax.

**Comment by Bill-McC** [comment]
@AdamSpeight2008 Surely the same visual issue applies to code such as :
Public Class Person
 Public Property FamilyName as String
 Public Property FirstName as String
 Public Property DOB as Date
End Class

If so then the suggestion should be for a generalised #begin #end block, eg:
#begin Public Property
     FamilyName As String
     FirstName As String
     DOB as Date
#end Public Property

That I could see useful for a wider range such as grouping of private vs pubic properties, methods etc, and yes, constants. But it wouldn't include type

**Comment by Bill-McC** [comment]
@scottdorman Yep, I was saying though the argument against enums casting doesn't stack up as that is already the rule. Another syntax like you suggest has issues that it appears to be a derivative of String which breaks fundamental rules about String.  I'm guessing by the : you meant Inherits in VB (BTW: it would make a lot more sense to suggest VB code in the VBLang group). Or would the : have another name ?

**Comment by scottdorman** [comment]
So, my example using VB syntax would be

```vb
Public Const ClaimTypes As String
   Expiration = "http://schemas.microsoft.com/ws/2008/06/identity/claims/expiration"
   Role = "http://schemas.microsoft.com/ws/2008/06/identity/claims/role"
   Anonymous = "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/anonymous"
   Country = "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/country"
   DateOfBirth = "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/dateofbirth"
   Email = "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress"
   Gender = "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/gender"
   GivenName = "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/givenname"
End Const
```

instead of 

```vb
Public Static Class ClaimTypes
   Public Const Expiration As String = "http://schemas.microsoft.com/ws/2008/06/identity/claims/expiration";
   Public Const Role As String = "http://schemas.microsoft.com/ws/2008/06/identity/claims/role";
   Public Const Anonymous As String = "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/anonymous";
   Public Const Country As String = "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/country";
   Public Const DateOfBirth As String = "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/dateofbirth";
   Public Const Email As String = "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress";
   Public Const Gender As String = "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/gender";
   Public Const GivenName As String = "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/givenname";
End Class
```

It still ends up being much more concise, compact and readable.

**Comment by AdamSpeight2008** [comment]
@Bill-McC  Would you mind putting section of code inside the code blocks so we differentiate code from text.

**Comment by AdamSpeight2008** [comment]
```vbnet
Public Const ClaimTypes As String
End Const
```
I ruled out as the header for the block `Public Const ClaimTypes As String` can be interpreted as an incomplete `Const Statement`.

**Comment by scottdorman** [comment]
@AdamSpeight2008 True, which simply means using `Const` as the keyword here might not be the right thing. I suppose another way to do it would be 

```vb
Public ClaimTypes As Const String
End
```
(although I don't know what the corresponding `End` keyword should be here)

which disambiguates it from an incomplete `Const` statement while still keeping the intent of a named group of string constants.

**Comment by AdamSpeight2008** [comment]
Secondly the proposal is only `syntactic sugar` for writing the existing code, **not** for a new code construct. The dotting into `ClaimTypes.` can in most case can be achieved via existing constructs, like inner class and modules etc. It doesn't cover the case of a block of constants within a method.

**Comment by AdamSpeight2008** [comment]
@KathleenDollard  @AnthonyDGreen
I am working on a prototype of this language feature over on my fork [Prototypes/ConstantBlock](https://github.com/AdamSpeight2008/roslyn-AdamSpeight2008/tree/Prototypes/ConstantBlock)
